src/core/Target_location.ml

```diff
@@ -0,0 +1,82 @@
+(* Cooper Pierce
+ *
+ * Copyright (C) 2023 Semgrep Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+type t = Code of code | Lockfile of lockfile [@@deriving show]
```

Comment:
Historically this would have only been one type, currently the `code` type (as `t` instead). However, with @mmcqd's recent changes in https://github.com/semgrep/semgrep/pull/9695, `Input_to_core.target` is now a sum in the ATD. So this is defined similarly by analogy (and to reduce the complexity of when I had rebased my changes on that PR to resolve conflicts).

I can change it so that `Target_location` unifies shared fields, but I was a bit hesitant to do so since I don't know what Matthew is relying on for his changes to pull supply chain into core.

Reply:
Unifying the shared fields seems like a good idea to me. The description of what these things are is in `Input_to_core.atd`. It definitely kinda feels like we're ending up with too many different types that are all very similar in name and structure, but with slightly different amounts of information.

Reply:
> It definitely kinda feels like we're ending up with too many different types that are all very similar in name and structure, but with slightly different amounts of information.

Yes, I think ideally we would just have one set of these, instead of two, split across here and `Input_to_core`, but while we still need to allow pysemgrep to send the existing list of targets it seems difficult to ditch `Input_to_core` in favour of non-ATD types.

Reply:
Wrt unifying the shared fields, I do think it makes sense, but I don't see a straightforward way to do so, since we rely on being able to deal with only `Target_location.code` values in some circumstances, and having this type be a record rather than a sum would remove that capability (at least without a lot of type mangaling at those use sites).

> I'd then like us to have separate modules for the fields for the other products so that Target_location can be generic.

It does sound like there is potentially some confusion that this is related to any particular product. That isn't the intention. Indeed, a supply chain or secrets-only scan would still have "code" targets. This is because the "code" part is not a description of the product, but rather of the sort of thing the target _contains_. Code targets are "code" since they comprise source code; not because they are used with the code/sast product.

Reply:
Since I'm sure I'm mentioning this for the first time, I hate the name of the Code product

Reply:
> we rely on being able to deal with only Target_location.code values in some circumstances

That's fair. Is it reasonable to do something like this?
```
type 'a target = {
   source: Source.t
   internal_path_to_content: Fpath.t
   details: 'a
}
```